main.c

```diff
@@ -685,6 +688,16 @@ static void *tcmur_cmdproc_thread(void *arg)
 			tcmu_err("ppoll received unexpected revent: 0x%x\n", pfd.revents);
 			break;
 		}
+
+		/*
+		 * If a device being removed is dealing with contiguous I/O requests,
+		 * we need to throttle cmdproc thread to make sure it have a chance
+		 * to become idle
+		 */
+		pthread_mutex_lock(&rdev->state_lock);
+		if (rdev->flags & TCMUR_DEV_FLAG_STOPPING)
+			dev_stopping = true;
+		pthread_mutex_unlock(&rdev->state_lock);
 	}
 
 	tcmu_err("thread terminating, should never happen\n");
```

Comment:
We can now hit this path right, so we should remove this.

Reply:
It looks like we can now hit the following:

We go down this code path and call pthread_cleanup_pop -> cmdproc_thread_cleanup -> rhandler->close

while commands are still being executed by the handler. Maybe we want the call to track_aio_request_finish in cmdproc_thread_cleanup instead of below in dev_removed.

Reply:
In consideration of that we already have the cmd_time_out fired in kernel to make sure outgoing IOs be handled, there is no need to drain the cmd ring buffer, right? 
add track_aio_request_finish logic in cmdproc_thread_cleanup make things kind of complicated, we need to distinguish whether handler has really take over outgoing IO.

Reply:
> @mikechristie In consideration of that we already have the cmd_time_out fired in kernel to make sure outgoing IOs be handled, there is no need to drain the cmd ring buffer, right?

For device removal, there is no need.

Later, when we fix up support for stopping and restarting tcmu-runner while devices are still present, then it could be more complicated. Would we want to just let the kernel cmd_time_out fire or loop over the ring buffer and fail commands or execute the commands? We can handle that later, so we can fix this crash now.


> add track_aio_request_finish logic in cmdproc_thread_cleanup make things kind of complicated, we need to distinguish whether handler has really take over outgoing IO.

Ok. Sounds good. Do the simple option now.

Reply:
For this code path were you going to drop the code where we can exit from the loop so we have a single place that can call cmdproc_thread_cleanup?

So we do not call cmdproc_thread_cleanup from 

tcmur_cmdproc_thread->pthread_cleanup_pop->cmdproc_thread_cleanup

and we were only going to call cmdproc_thread_cleanup from 

dev_removed->tcmulib_cleanup_cmdproc_thread->cancel_thread->pthread_cancel->cmdproc_thread_cleanup 

right?

Reply:
I saw you repushed your patches. Were you going to handle this too? Right now wen can do tcmur_cmdproc_thread->pthread_cleanup_pop->cmdproc_thread_cleanup->rhandler->close which could free memory that the handler is using to execute io. I think we can just do the attached patch.

[only-execute-cmdproc-cleanup-once.txt](https://github.com/open-iscsi/tcmu-runner/files/1296877/only-execute-cmdproc-cleanup-once.txt)

Reply:
@mikechristie Sorry for missing your comments here. @lxbsz Thanks for your reminding:-) Yes, you are right about cmdproc_thread_cleanup, repush this PR again with your patch merged.